### PR TITLE
fix: validate duplicate programRule name

### PR DESCRIPTION
**Pull request by vietnguyen**
https://dhis2.atlassian.net/browse/DHIS2-13502

### CODE REVIEW

**Review by enricocolasante — APPROVED** [review]
Do we need a script or something else to fix or spot duplication already present in the DB?

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/ProgramRuleControllerTest.java [hunk 1]
```diff
@@ -57,4 +60,20 @@ void testValidateCondition_NoSuchProgram()
...
+        manager.save( program );
+        assertStatus( HttpStatus.OK, GET( "/programs/{id}", program.getUid() ) );
+        assertStatus( HttpStatus.OK,
+            POST( "/metadata/", "{'programRules':[{'name':'test', 'program':{ 'id':'" + program.getUid() + "'}}]}" ) );
+
+        JsonImportSummary response = POST( "/metadata/",
+            "{'programRules':[{'name':'test', 'program':{ 'id':'" + program.getUid() + "'}}]}" )
+                .content( HttpStatus.CONFLICT ).get( "response" ).as( JsonImportSummary.class );
+        assertEquals( "The Program Rule name test already exist in Program ProgramA",
+            response.getTypeReports().get( 0 ).getObjectReports().get( 0 ).getErrorReports().get( 0 ).getMessage() );
```
Comment (jbee) [inline dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/ProgramRuleControllerTest.java:77]: FYI: Digging down to a particular report might become unstable should more conflicts be found when new validations are added. There are `find` methods inherited from `JsonObject` where you can essentially grab a conflict by the error code and then check its message. I have done this in some tests.

**Review by jbee — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+import org.hisp.dhis.dxf2.metadata.objectbundle.ObjectBundle;
+import org.hisp.dhis.feedback.ErrorCode;
+import org.hisp.dhis.feedback.ErrorReport;
+import org.hisp.dhis.program.Program;
+import org.hisp.dhis.programrule.ProgramRule;
+import org.hisp.dhis.programrule.ProgramRuleService;
+import org.springframework.stereotype.Component;
+
+@Component
+@AllArgsConstructor
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java:44]: Maybe better to use `@RequiredArgsConstructor`

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        Program program = programRule.getProgram();
+
+        if ( program == null )
+        {
+            return;
+        }
+
+        program = bundle.getPreheat().get( bundle.getPreheatIdentifier(), program );
+
+        List<ProgramRule> rules = programRuleService.getProgramRule( program );
```
Comment (mortenoh) [inline dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java:61]: Should program rules also be part of the preheat?

File: dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/ProgramRuleControllerTest.java (hunk 1 above)
Comment (vietnguyen) [reply dhis-2/dhis-web-api-test/src/test/java/org/hisp/dhis/webapi/controller/ProgramRuleControllerTest.java]: fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java (hunk 2 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java]: fixed

File: dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java (hunk 3 above)
Comment (vietnguyen) [reply dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/metadata/objectbundle/hooks/ProgramRuleObjectBundleHook.java:61]: I don't think so, by design `Program` object  doesn't have a collection of `ProgramRule`.
If you have only one `ProgramRule` in payload, then preheat will only contain the referenced `Program` object.

**Comment by vietnguyen** [comment]
> Do we need a script or something else to fix or spot duplication already present in the DB?

@enricocolasante since we are not backporting this PR and this change won't break any function in program rule engine, I'm thinking just adding it to the 2.40 Upgrade Notes document. So user will be aware of the change and update duplicate names accordingly when they get the error message in Maintenance app.

**Pull request merged**